File: src/main/java/ru/ezhov/rocket/action/configuration/ui/ConfigurationFrame.java
package ru.ezhov.rocket.action.configuration.ui;

import ru.ezhov.rocket.action.RocketActionSettingsRepository;
import ru.ezhov.rocket.action.RocketActionSettingsRepositoryException;
import ru.ezhov.rocket.action.RocketActionUiRepository;
import ru.ezhov.rocket.action.api.RocketActionConfiguration;
import ru.ezhov.rocket.action.api.RocketActionConfigurationProperty;
import ru.ezhov.rocket.action.api.RocketActionSettings;
import ru.ezhov.rocket.action.icon.AppIcon;
import ru.ezhov.rocket.action.icon.IconRepositoryFactory;
import ru.ezhov.rocket.action.infrastructure.MutableRocketActionSettings;
import ru.ezhov.rocket.action.notification.NotificationFactory;
import ru.ezhov.rocket.action.notification.NotificationType;

import javax.swing.AbstractAction;
import javax.swing.Action;
import javax.swing.DropMode;
import javax.swing.JButton;
import javax.swing.JDialog;
import javax.swing.JLabel;
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;
import javax.swing.JPanel;
import javax.swing.JPopupMenu;
import javax.swing.JScrollPane;
import javax.swing.JSplitPane;
import javax.swing.JTable;
import javax.swing.JTree;
import javax.swing.SwingUtilities;
import javax.swing.table.DefaultTableModel;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeCellRenderer;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.MutableTreeNode;
import javax.swing.tree.TreePath;
import javax.swing.tree.TreeSelectionModel;
import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.Dialog;
import java.awt.Dimension;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;
import java.util.Vector;
import java.util.stream.Collectors;

public class ConfigurationFrame {

    private JDialog dialog;
    private RocketActionConfigurationRepository rocketActionConfigurationRepository;
    private RocketActionUiRepository rocketActionUiRepository;
    private CreateRocketActionSettingsDialog createRocketActionSettingsDialog;
    private RocketActionSettingsRepository rocketActionSettingsRepository;
    private ActionListener updateActionListener;

    public ConfigurationFrame(
            Dialog owner,
            RocketActionConfigurationRepository rocketActionConfigurationRepository,
            RocketActionUiRepository rocketActionUiRepository,
            RocketActionSettingsRepository rocketActionSettingsRepository,
            ActionListener updateActionListener
    ) throws Exception {
        dialog = new JDialog(owner, "Rocket action configuration");

        this.updateActionListener = updateActionListener;
        this.rocketActionConfigurationRepository = rocketActionConfigurationRepository;
        this.rocketActionUiRepository = rocketActionUiRepository;
        this.rocketActionSettingsRepository = rocketActionSettingsRepository;

        JMenuBar menuBar = new JMenuBar();

        JMenuItem menuItemUpdate = new JMenuItem("Update");
        menuItemUpdate.setIcon(IconRepositoryFactory.instance().by(AppIcon.RELOAD));
        menuItemUpdate.addActionListener(e -> SwingUtilities.invokeLater(() -> {
            updateActionListener.actionPerformed(e);
            ConfigurationFrame.this.setVisible(false);
        }));
        menuBar.add(menuItemUpdate);


        final Dimension size = Toolkit.getDefaultToolkit().getScreenSize();
        dialog.setSize((int) (size.width * 0.6), (int) (size.height * 0.6));
        dialog.setLocationRelativeTo(null);
        createRocketActionSettingsDialog = new CreateRocketActionSettingsDialog(
                dialog,
                rocketActionConfigurationRepository,
                rocketActionUiRepository
        );
        dialog.add(menuBar, BorderLayout.NORTH);
        dialog.add(panel(), BorderLayout.CENTER);
    }

    private JPanel panel() throws Exception {
        JPanel panel = new JPanel(new BorderLayout());
        panel.add(tree(), BorderLayout.CENTER);
        return panel;
    }

    private JPanel tree() throws Exception {
        JSplitPane splitPane = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT);
        splitPane.setDividerLocation(0.5D);
        splitPane.setResizeWeight(0.5D);

        final List<RocketActionSettings> actions = rocketActionSettingsRepository.actions();

        DefaultMutableTreeNode root = new DefaultMutableTreeNode(null, true);
        fillTreeNodes(actions, root);
        DefaultTreeModel defaultTreeModel = new DefaultTreeModel(root);

        RocketActionSettingsPanel rocketActionSettingsPanel = new RocketActionSettingsPanel();

        JPanel panel = new JPanel(new BorderLayout());
        JTree tree = new JTree(defaultTreeModel);
        tree.getSelectionModel().setSelectionMode(TreeSelectionModel.SINGLE_TREE_SELECTION);
        tree.addTreeSelectionListener(e -> {
            final TreePath path = e.getNewLeadSelectionPath();
            if (path == null) return;
            DefaultMutableTreeNode node = (DefaultMutableTreeNode) path.getLastPathComponent();

            if (node == null) return;

            Object o = node.getUserObject();
            if (o != null) {
                RocketActionSettings settings = (RocketActionSettings) o;
                SwingUtilities.invokeLater(
                        () -> rocketActionSettingsPanel.show(
                                settings,
                                node::setUserObject
                        )
                );
            }
        });
        tree.setCellRenderer(new RocketActionSettingsCellRender());
        tree.setRootVisible(false);

        JPanel panelTree = new JPanel(new BorderLayout());
        panelTree.add(new JScrollPane(tree), BorderLayout.CENTER);
        JPanel panelSaveTree = new JPanel();

        JButton buttonSaveTree = new JButton("Save actions");
        buttonSaveTree.addActionListener(e -> saveSettings(defaultTreeModel));

        panelSaveTree.add(buttonSaveTree);

        panelTree.add(panelSaveTree, BorderLayout.SOUTH);

        splitPane.setLeftComponent(panelTree);
        splitPane.setRightComponent(rocketActionSettingsPanel);

        tree.setDragEnabled(true);
        tree.setDropMode(DropMode.ON_OR_INSERT);
        tree.setTransferHandler(new TreeTransferHandler());

        tree.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseReleased(MouseEvent e) {
                if (e.getButton() == MouseEvent.BUTTON3) {
                    final TreePath treePath = tree.getClosestPathForLocation(e.getX(), e.getY());
                    if (treePath == null) return;

                    DefaultMutableTreeNode mutableTreeNode = (DefaultMutableTreeNode) treePath.getLastPathComponent();

                    JPopupMenu popupMenu = new JPopupMenu();
                    popupMenu.add(new JMenuItem(new AbstractAction() {
                        {
                            putValue(Action.NAME, "Add new TOP");
                        }

                        @Override
                        public void actionPerformed(ActionEvent e) {
                            createRocketActionSettingsDialog.show(rocketActionSettings -> {
                                SwingUtilities.invokeLater(() -> {
                                    MutableRocketActionSettings newActionSettings = new MutableRocketActionSettings(
                                            rocketActionSettings.id(),
                                            rocketActionSettings.type(),
                                            rocketActionSettings.settings(),
                                            rocketActionSettings.actions()
                                    );

                                    defaultTreeModel.insertNodeInto(
                                            new DefaultMutableTreeNode(newActionSettings, true),
                                            (MutableTreeNode) mutableTreeNode.getParent(),
                                            mutableTreeNode.getParent().getIndex(mutableTreeNode)

                                    );
                                });
                            });
                        }
                    }));
                    popupMenu.add(new JMenuItem(
                            new AbstractAction() {
                                {
                                    putValue(Action.NAME, "Add new DOWN");
                                }

                                @Override
                                public void actionPerformed(ActionEvent e) {
                                    createRocketActionSettingsDialog.show(rocketActionSettings -> {
                                        SwingUtilities.invokeLater(() -> {
                                            defaultTreeModel.insertNodeInto(
                                                    new DefaultMutableTreeNode(
                                                            new MutableRocketActionSettings(
                                                                    rocketActionSettings.id(),
                                                                    rocketActionSettings.type(),
                                                                    rocketActionSettings.settings(),
                                                                    rocketActionSettings.actions()
                                                            ),
                                                            true
                                                    ),
                                                    (MutableTreeNode) mutableTreeNode.getParent(),
                                                    mutableTreeNode.getParent().getIndex(mutableTreeNode) + 1
                                            );
                                        });
                                    });
                                }
                            }
                    ));
                    popupMenu.add(new JMenuItem(
                            new AbstractAction() {
                                {
                                    putValue(Action.NAME, "Add new as child");
                                }

                                @Override
                                public void actionPerformed(ActionEvent e) {
                                    createRocketActionSettingsDialog.show(rocketActionSettings -> {
                                        SwingUtilities.invokeLater(() -> {
                                            mutableTreeNode.add(
                                                    new DefaultMutableTreeNode(
                                                            new MutableRocketActionSettings(
                                                                    rocketActionSettings.id(),
                                                                    rocketActionSettings.type(),
                                                                    rocketActionSettings.settings(),
                                                                    rocketActionSettings.actions()
                                                            ),
                                                            true
                                                    )
                                            );

                                            defaultTreeModel.reload(mutableTreeNode);
                                        });
                                    });
                                }
                            }
                    ));

                    popupMenu.add(new JMenuItem(
                            new AbstractAction() {
                                {
                                    putValue(Action.NAME, "Delete");
                                }

                                @Override
                                public void actionPerformed(ActionEvent e) {
                                    SwingUtilities.invokeLater(() -> {
                                        mutableTreeNode.removeFromParent();
                                        defaultTreeModel.reload();
                                    });
                                }
                            }
                    ));

                    popupMenu.show(e.getComponent(), e.getX(), e.getY());
                }
            }
        });

        panel.add(splitPane, BorderLayout.CENTER);
        return panel;
    }

    private void fillTreeNodes(List<RocketActionSettings> actions, DefaultMutableTreeNode parent) {
        for (RocketActionSettings rocketActionSettings : actions) {
            DefaultMutableTreeNode current = new DefaultMutableTreeNode(rocketActionSettings, true);
            parent.add(current);
            if (!rocketActionSettings.actions().isEmpty()) {
                final List<RocketActionSettings> childAction = rocketActionSettings.actions();
                fillTreeNodes(childAction, current);
            }
        }
    }

    private void saveSettings(DefaultTreeModel treeModel) {
        List<RocketActionSettings> settings = new ArrayList<>();
        DefaultMutableTreeNode root = (DefaultMutableTreeNode) treeModel.getRoot();
        int childCount = root.getChildCount();
        for (int i = 0; i < childCount; i++) {
            recursiveGetSettings((DefaultMutableTreeNode) root.getChildAt(i), settings, null);
        }

        try {
            rocketActionSettingsRepository.save(settings);

            NotificationFactory.getInstance().show(NotificationType.INFO, "Actions saved");
        } catch (RocketActionSettingsRepositoryException e) {
            e.printStackTrace();

            NotificationFactory.getInstance().show(NotificationType.ERROR, "Error actions saving");
        }
    }

    private void recursiveGetSettings(DefaultMutableTreeNode node, List<RocketActionSettings> settings, MutableRocketActionSettings parent) {
        MutableRocketActionSettings originalActionSettings = (MutableRocketActionSettings) node.getUserObject();
        MutableRocketActionSettings finalActionSettings =
                new MutableRocketActionSettings(
                        originalActionSettings.id(),
                        originalActionSettings.type(),
                        originalActionSettings.settings()
                );
        if (parent == null) {
            settings.add(finalActionSettings);
        } else {
            parent.add(finalActionSettings);
        }
        int childCount = node.getChildCount();
        for (int i = 0; i < childCount; i++) {
            recursiveGetSettings((DefaultMutableTreeNode) node.getChildAt(i), settings, finalActionSettings);
        }
    }

    public void setVisible(boolean visible) {
        dialog.setVisible(visible);
    }

    private class RocketActionSettingsCellRender extends DefaultTreeCellRenderer {
        @Override
        public Component getTreeCellRendererComponent(JTree tree, Object value, boolean sel, boolean expanded, boolean leaf, int row, boolean hasFocus) {
            JLabel label = (JLabel) super.getTreeCellRendererComponent(tree, value, sel, expanded, leaf, row, hasFocus);

            if (value != null) {
                DefaultMutableTreeNode node = (DefaultMutableTreeNode) value;
                if (node.getUserObject() instanceof RocketActionSettings) {
                    RocketActionSettings settings = (RocketActionSettings) node.getUserObject();
                    if (settings != null) {
                        final String labelProperty = settings.settings().get("label");
                        if (labelProperty != null && !"".equals(labelProperty)) {
                            label.setText(labelProperty);
                        } else {
                            label.setText(settings.type());
                        }
                    }
                }
            }

            return label;
        }
    }

    private class RocketActionSettingsPanel extends JPanel {
        private final DefaultTableModel tableModel = new DefaultTableModel();
        private final JTable table = new JTable(tableModel);
        private RocketActionSettings currentSettings;
        private SavedRocketActionSettingsPanelCallback callback;
        private JLabel labelType = new JLabel();
        private JLabel labelDescription = new JLabel();

        public RocketActionSettingsPanel() {
            super(new BorderLayout());
            tableModel.addColumn("Name");
            tableModel.addColumn("Value");
            tableModel.addColumn("Description");
            add(top(), BorderLayout.NORTH);
            add(new JScrollPane(table), BorderLayout.CENTER);
            add(testAndCreate(), BorderLayout.SOUTH);
        }

        private JPanel top() {
            JPanel panel = new JPanel(new BorderLayout());
            panel.add(labelType, BorderLayout.NORTH);
            panel.add(labelDescription, BorderLayout.CENTER);
            return panel;
        }

        private JPanel testAndCreate() {
            JPanel panel = new JPanel();
            JButton button = new JButton("Save current action");
            button.addActionListener(e -> {
                callback.saved(create());
                NotificationFactory.getInstance().show(NotificationType.INFO, "Current action saved");
            });
            panel.add(button);
            return panel;
        }

        public void show(RocketActionSettings settings, SavedRocketActionSettingsPanelCallback callback) {
            this.currentSettings = settings;

            labelType.setText(settings.type());

            this.callback = callback;
            while (tableModel.getRowCount() != 0) {
                tableModel.removeRow(0);
            }

            final Optional<RocketActionConfiguration> configurationOptional = rocketActionConfigurationRepository.by(settings.type());

            RocketActionConfiguration configuration = null;
            if (configurationOptional.isPresent()) {
                configuration = configurationOptional.get();

                labelDescription.setText(configuration.description());
            }

            final Map<String, String> map = settings.settings();
            RocketActionConfiguration finalConfiguration = configuration;
            map.forEach((k, v) -> {
                Vector<String> row = new Vector<>();
                row.add(k);
                row.add(v);
                if (finalConfiguration != null) {
                    final Optional<RocketActionConfigurationProperty> optional =
                            finalConfiguration
                                    .properties()
                                    .stream()
                                    .filter(p -> p.name().equals(k))
                                    .findFirst();
                    if (optional.isPresent()) {
                        row.add(optional.get().description());
                    } else {
                        row.add("Unregistered property");
                    }
                } else {
                    row.add(null);
                }
                tableModel.addRow(row);
            });

            final List<RocketActionConfigurationProperty> notSetProperty =
                    finalConfiguration
                            .properties()
                            .stream()
                            .filter(p -> !map.containsKey(p.name()))
                            .collect(Collectors.toList());

            notSetProperty.forEach(p -> {
                Vector<String> row = new Vector<>();
                row.add(p.name());
                row.add("");
                row.add(p.description());
                tableModel.addRow(row);
            });
        }

        private RocketActionSettings create() {
            if (currentSettings == null)
                throw new IllegalStateException("Must be set current selected configuration");

            int rowCount = tableModel.getRowCount();
            Map<String, String> map = new TreeMap<>();
            for (int i = 0; i < rowCount; i++) {
                Object name = tableModel.getValueAt(i, 0);
                Object value = tableModel.getValueAt(i, 1);
                map.put(name.toString(), value.toString());
            }

            return new MutableRocketActionSettings(
                    currentSettings.id(),
                    currentSettings.type(),
                    map,
                    currentSettings.actions()
            );
        }
    }
}

File: src/main/kotlin/ru/ezhov/rocket/action/configuration/ui/ConfigurationFrame.kt
package ru.ezhov.rocket.action.configuration.ui

import mu.KotlinLogging
import ru.ezhov.rocket.action.api.RocketActionSettings
import ru.ezhov.rocket.action.configuration.domain.RocketActionConfigurationRepository
import ru.ezhov.rocket.action.domain.RocketActionSettingsRepository
import ru.ezhov.rocket.action.domain.RocketActionSettingsRepositoryException
import ru.ezhov.rocket.action.domain.RocketActionUiRepository
import ru.ezhov.rocket.action.icon.AppIcon
import ru.ezhov.rocket.action.icon.IconRepositoryFactory
import ru.ezhov.rocket.action.infrastructure.MutableRocketActionSettings
import ru.ezhov.rocket.action.notification.NotificationFactory
import ru.ezhov.rocket.action.notification.NotificationType
import ru.ezhov.rocket.action.types.group.GroupRocketActionUi
import java.awt.BorderLayout
import java.awt.Component
import java.awt.Dialog
import java.awt.Toolkit
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.AbstractAction
import javax.swing.DropMode
import javax.swing.JButton
import javax.swing.JDialog
import javax.swing.JLabel
import javax.swing.JMenuBar
import javax.swing.JMenuItem
import javax.swing.JPanel
import javax.swing.JPopupMenu
import javax.swing.JScrollPane
import javax.swing.JSplitPane
import javax.swing.JTree
import javax.swing.SwingUtilities
import javax.swing.event.TreeExpansionEvent
import javax.swing.event.TreeExpansionListener
import javax.swing.event.TreeSelectionEvent
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.DefaultTreeCellRenderer
import javax.swing.tree.DefaultTreeModel
import javax.swing.tree.MutableTreeNode
import javax.swing.tree.TreePath
import javax.swing.tree.TreeSelectionModel

private val logger = KotlinLogging.logger { }

class ConfigurationFrame(
        owner: Dialog,
        rocketActionConfigurationRepository: RocketActionConfigurationRepository,
        rocketActionUiRepository: RocketActionUiRepository,
        rocketActionSettingsRepository: RocketActionSettingsRepository?,
        updateActionListener: ActionListener
) {
    private val dialog: JDialog
    private val rocketActionConfigurationRepository: RocketActionConfigurationRepository
    private val rocketActionUiRepository: RocketActionUiRepository
    private val createRocketActionSettingsDialog: CreateRocketActionSettingsDialog
    private val rocketActionSettingsRepository: RocketActionSettingsRepository?
    private val updateActionListener: ActionListener

    @Throws(Exception::class)
    private fun panel(): JPanel {
        val panel = JPanel(BorderLayout())
        panel.add(tree(), BorderLayout.CENTER)
        return panel
    }

    @Throws(Exception::class)
    private fun tree(): JPanel {
        val splitPane = JSplitPane(JSplitPane.HORIZONTAL_SPLIT)
        splitPane.setDividerLocation(0.5)
        splitPane.resizeWeight = 0.5
        val actions = rocketActionSettingsRepository!!.actions()
        val root = DefaultMutableTreeNode(null, true)
        fillTreeNodes(actions, root)
        val defaultTreeModel = DefaultTreeModel(root)
        val rocketActionSettingsPanel = EditorRocketActionSettingsPanel(
                rocketActionConfigurationRepository,
                rocketActionUiRepository
        )
        val panel = JPanel(BorderLayout())
        val tree = JTree(defaultTreeModel)
        tree.selectionModel.selectionMode = TreeSelectionModel.SINGLE_TREE_SELECTION
        tree.addTreeSelectionListener { e: TreeSelectionEvent ->
            val path = e.newLeadSelectionPath ?: return@addTreeSelectionListener
            val node = path.lastPathComponent as DefaultMutableTreeNode
            val o = node.userObject
            if (o != null) {
                val settings = o as TreeRocketActionSettings
                SwingUtilities.invokeLater {
                    rocketActionSettingsPanel.show(
                            settings = settings,
                            callback = object : SavedRocketActionSettingsPanelCallback {
                                override fun saved(settings: TreeRocketActionSettings) {
                                    node.userObject = settings
                                }
                            }
                    )
                }
            }
        }
        tree.cellRenderer = RocketActionSettingsCellRender()
        tree.isRootVisible = false
        val panelTree = JPanel(BorderLayout())
        panelTree.add(JScrollPane(tree), BorderLayout.CENTER)
        val panelSaveTree = JPanel()
        val buttonSaveTree = JButton("Сохранить всю конфигурацию")
        buttonSaveTree.addActionListener { saveSettings(defaultTreeModel) }
        panelSaveTree.add(buttonSaveTree)
        panelTree.add(panelSaveTree, BorderLayout.SOUTH)
        splitPane.leftComponent = panelTree
        splitPane.rightComponent = rocketActionSettingsPanel
        tree.dragEnabled = true
        tree.dropMode = DropMode.ON_OR_INSERT
        tree.transferHandler = TreeTransferHandler()
        val expandedSet = mutableSetOf<TreePath>()
        tree.addTreeExpansionListener(object : TreeExpansionListener {
            override fun treeExpanded(event: TreeExpansionEvent?) {
                event?.let {
                    logger.debug { "Expanded tree path '${it.path.lastPathComponent}'" }
                    expandedSet.add(it.path)
                }
            }

            override fun treeCollapsed(event: TreeExpansionEvent?) {
                event?.let {
                    logger.debug { "Collapsed tree path '${it.path.lastPathComponent}'" }
                    expandedSet.remove(it.path)
                }
            }
        })
        tree.addMouseListener(object : MouseAdapter() {
            override fun mouseReleased(e: MouseEvent) {
                if (e.button == MouseEvent.BUTTON3) {
                    val treePath = tree.getClosestPathForLocation(e.x, e.y) ?: return
                    val mutableTreeNode = treePath.lastPathComponent as DefaultMutableTreeNode
                    val userObject = mutableTreeNode.userObject as? RocketActionSettings
                    val popupMenu = JPopupMenu()
                    popupMenu.add(JMenuItem(object : AbstractAction() {
                        override fun actionPerformed(e: ActionEvent) {
                            createRocketActionSettingsDialog.show(object : CreatedRocketActionSettingsCallback {
                                override fun create(rocketActionSettings: RocketActionSettings) {
                                    SwingUtilities.invokeLater {
                                        val newActionSettings = MutableRocketActionSettings(
                                                rocketActionSettings.id(),
                                                rocketActionSettings.type(),
                                                rocketActionSettings.settings().toMutableMap(),
                                                rocketActionSettings.actions().toMutableList()
                                        )
                                        defaultTreeModel.insertNodeInto(
                                                DefaultMutableTreeNode(newActionSettings, true),
                                                mutableTreeNode.parent as MutableTreeNode,
                                                mutableTreeNode.parent.getIndex(mutableTreeNode)
                                        )
                                    }
                                }
                            })
                        }

                        init {
                            putValue(NAME, "Добавить выше")
                        }
                    }))
                    popupMenu.add(JMenuItem(
                            object : AbstractAction() {
                                override fun actionPerformed(e: ActionEvent) {
                                    createRocketActionSettingsDialog.show(object : CreatedRocketActionSettingsCallback {
                                        override fun create(rocketActionSettings: RocketActionSettings) {
                                            SwingUtilities.invokeLater {
                                                defaultTreeModel.insertNodeInto(
                                                        DefaultMutableTreeNode(
                                                                MutableRocketActionSettings(
                                                                        rocketActionSettings.id(),
                                                                        rocketActionSettings.type(),
                                                                        rocketActionSettings.settings().toMutableMap(),
                                                                        rocketActionSettings.actions().toMutableList()
                                                                ),
                                                                true
                                                        ),
                                                        mutableTreeNode.parent as MutableTreeNode,
                                                        mutableTreeNode.parent.getIndex(mutableTreeNode) + 1
                                                )
                                            }
                                        }

                                    })
                                }

                                init {
                                    putValue(NAME, "Добавить ниже")
                                }
                            }
                    ))
                    if (userObject?.type()?.value() == GroupRocketActionUi.TYPE) {
                        popupMenu.add(JMenuItem(
                                object : AbstractAction() {
                                    override fun actionPerformed(e: ActionEvent) {
                                        createRocketActionSettingsDialog.show(object : CreatedRocketActionSettingsCallback {
                                            override fun create(rocketActionSettings: RocketActionSettings) {
                                                SwingUtilities.invokeLater {
                                                    mutableTreeNode.add(
                                                            DefaultMutableTreeNode(
                                                                    MutableRocketActionSettings(
                                                                            rocketActionSettings.id(),
                                                                            rocketActionSettings.type(),
                                                                            rocketActionSettings.settings().toMutableMap(),
                                                                            rocketActionSettings.actions().toMutableList()
                                                                    ),
                                                                    true
                                                            )
                                                    )
                                                    defaultTreeModel.reload(mutableTreeNode)
                                                }
                                            }
                                        })
                                    }

                                    init {
                                        putValue(NAME, "Создать и добавить как потомка")
                                    }
                                }
                        ))
                    }
                    popupMenu.add(JMenuItem(
                            object : AbstractAction() {
                                override fun actionPerformed(e: ActionEvent) {
                                    SwingUtilities.invokeLater {
                                        val parent = mutableTreeNode.parent
                                        mutableTreeNode.removeFromParent()
                                        defaultTreeModel.reload(parent)

                                        expandedSet.forEach { path ->
                                            tree.expandPath(path)
                                        }
                                    }
                                }

                                init {
                                    putValue(NAME, "Удалить")
                                }
                            }
                    ))
                    popupMenu.show(e.component, e.x, e.y)
                }
            }
        })
        panel.add(splitPane, BorderLayout.CENTER)
        return panel
    }

    private fun fillTreeNodes(actions: List<RocketActionSettings>?, parent: DefaultMutableTreeNode) {
        for (rocketActionSettings in actions!!) {
            rocketActionConfigurationRepository.by(rocketActionSettings.type())?.let { config ->
                val current = DefaultMutableTreeNode(
                        TreeRocketActionSettings(
                                config,
                                rocketActionSettings,
                        ),
                        true
                )
                parent.add(current)
                if (rocketActionSettings.actions().isNotEmpty()) {
                    val childAction = rocketActionSettings.actions()
                    fillTreeNodes(childAction, current)
                }
            } ?: run {
                logger.warn { "Configuration for settings '${rocketActionSettings.type()}' not found and skipped" }
            }
        }
    }

    private fun saveSettings(treeModel: DefaultTreeModel) {
        val settings: MutableList<RocketActionSettings> = ArrayList()
        val root = treeModel.root as DefaultMutableTreeNode
        val childCount = root.childCount
        for (i in 0 until childCount) {
            recursiveGetSettings(root.getChildAt(i) as DefaultMutableTreeNode, settings, null)
        }
        try {
            rocketActionSettingsRepository!!.save(settings)
            NotificationFactory.notification.show(NotificationType.INFO, "Действия сохранены")
        } catch (e: RocketActionSettingsRepositoryException) {
            e.printStackTrace()
            NotificationFactory.notification.show(NotificationType.ERROR, "Ошибка сохранения действий")
        }
    }

    private fun recursiveGetSettings(node: DefaultMutableTreeNode, settings: MutableList<RocketActionSettings>, parent: MutableRocketActionSettings?) {
        val originalActionSettings = node.userObject as MutableRocketActionSettings
        val finalActionSettings = MutableRocketActionSettings(
                originalActionSettings.id(),
                originalActionSettings.type(),
                originalActionSettings.settings()
        )
        parent?.add(finalActionSettings) ?: settings.add(finalActionSettings)
        val childCount = node.childCount
        for (i in 0 until childCount) {
            recursiveGetSettings(node.getChildAt(i) as DefaultMutableTreeNode, settings, finalActionSettings)
        }
    }

    fun setVisible(visible: Boolean) {
        dialog.isVisible = visible
    }

    private inner class RocketActionSettingsCellRender : DefaultTreeCellRenderer() {
        override fun getTreeCellRendererComponent(tree: JTree, value: Any, sel: Boolean, expanded: Boolean, leaf: Boolean, row: Int, hasFocus: Boolean): Component {
            val label = super.getTreeCellRendererComponent(tree, value, sel, expanded, leaf, row, hasFocus) as JLabel
            val node = value as DefaultMutableTreeNode
            if (node.userObject is TreeRocketActionSettings) {
                val settings = node.userObject as TreeRocketActionSettings
                label.text = settings.asString()
            }
            return label
        }
    }


    init {
        dialog = JDialog(owner, "Конфигурирование действий")
        this.updateActionListener = updateActionListener
        this.rocketActionConfigurationRepository = rocketActionConfigurationRepository
        this.rocketActionUiRepository = rocketActionUiRepository
        this.rocketActionSettingsRepository = rocketActionSettingsRepository
        val menuBar = JMenuBar()
        val menuItemUpdate = JMenuItem("Обновить")
        menuItemUpdate.icon = IconRepositoryFactory.repository.by(AppIcon.RELOAD)
        menuItemUpdate.addActionListener { e: ActionEvent? ->
            SwingUtilities.invokeLater {
                updateActionListener.actionPerformed(e)
                setVisible(false)
            }
        }
        menuBar.add(menuItemUpdate)
        val size = Toolkit.getDefaultToolkit().screenSize
        dialog.setSize((size.width * 0.6).toInt(), (size.height * 0.6).toInt())
        dialog.setLocationRelativeTo(null)
        createRocketActionSettingsDialog = CreateRocketActionSettingsDialog(
                dialog,
                rocketActionConfigurationRepository,
                rocketActionUiRepository
        )
        dialog.add(menuBar, BorderLayout.NORTH)
        dialog.add(panel(), BorderLayout.CENTER)
    }
}